python/pyarrow/dataset.py

```diff
@@ -153,11 +153,17 @@ def partitioning(schema=None, field_names=None, flavor=None):
 
 def _ensure_fs(filesystem, path):
     # Validate or infer the filesystem from the path
-    from pyarrow.fs import FileSystem
+    from pyarrow.fs import FileSystem, LocalFileSystem
 
-    if filesystem is not None:
-        return filesystem, path
-    return FileSystem.from_uri(path)
+    if filesystem is None:
+        try:
+            return FileSystem.from_uri(path)
+        except Exception:
```

Comment:
Shouldn't you be a bit more specific here? Some filesystems might fail instantiating for other reasons (e.g. bad S3 configuration?).

Reply:
Yeah, that's true (copied it from the old code without thinking about it much further).

I suppose the problem, though, is which error? Right now, this gives an ArrowInvalid:

```
In [31]: pyarrow.fs.FileSystem.from_uri('blabla')  
---------------------------------------------------------------------------
ArrowInvalid                              Traceback (most recent call last)
<ipython-input-31-64eb29ab5232> in <module>
----> 1 pyarrow.fs.FileSystem.from_uri('blabla')

~/scipy/repos/arrow/python/pyarrow/_fs.pyx in pyarrow._fs.FileSystem.from_uri()

~/scipy/repos/arrow/python/pyarrow/error.pxi in pyarrow.lib.pyarrow_internal_check_status()

~/scipy/repos/arrow/python/pyarrow/error.pxi in pyarrow.lib.check_status()

ArrowInvalid: URI has empty scheme: 'blabla'
```

Do you know what eg S3 gives?

Alternatively, I could also explicitly check first if the path exists as a (relative) local file path, instead of relying on catching an error to fall back

Reply:
Well, right now, it seems creating a filesystem can also give a `ArrowInvalid`, if some parameters are invalid (for example if you are giving a S3 scheme other than "http" or "https").

In any case, `except Exception` is wrong.

Reply:
OK, but so catching `ArrowInvalid` (although certainly better) won't solve the actual concern about other errors being catched, right?

So checking if the path exists as a local path might be better?

Reply:
> So checking if the path exists as a local path might be better?

Yes, that sounds reasonable. At least for _reading_ a dataset...